src/main/java/org/testng/internal/BaseInvoker.java

```diff
@@ -51,8 +54,16 @@ protected void runInvokedMethodListeners(
 
     InvokedMethodListenerInvoker invoker =
         new InvokedMethodListenerInvoker(listenerMethod, testResult, m_testContext);
-    for (IInvokedMethodListener currentListener : m_invokedMethodListeners) {
-      invoker.invokeListener(currentListener, invokedMethod);
+    if(InvokedMethodListenerMethod.BEFORE_INVOCATION.equals(listenerMethod)) {
```

Comment:
We could avoid the nesting by doing something like this 

```java
List<IInvokedMethodListener> m_invokedMethodListenerList = new ArrayList<>(m_invokedMethodListeners);
if(InvokedMethodListenerMethod.AFTER_INVOCATION.equals(listenerMethod)) {
    Collections.reverse(m_invokedMethodListenerList);
}

for (IInvokedMethodListener currentListener : m_invokedMethodListenerList) {
      invoker.invokeListener(currentListener, invokedMethod);
}
```
This does away with the `if..else` and duplicating the iteration in both the places.

Reply:
Will update it.